applications/FluidDynamicsApplication/custom_elements/data_containers/qs_vms/qs_vms_derivative_utilities.cpp

```diff
@@ -0,0 +1,413 @@
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/variables.h"
+#include "utilities/element_size_calculator.h"
+
+// Application includes
+
+// Include base h
+#include "qs_vms_derivative_utilities.h"
+
+namespace Kratos
+{
+namespace QSVMSDerivativeHelperUtilities
+{
+template<unsigned int TComponentIndex>
+const Variable<double>& GetVelocityVariable();
+
+template<unsigned int TComponentIndex>
+const Variable<double>& GetShapeVariable();
+
+template<>
+const Variable<double>& GetVelocityVariable<0>()
+{
+    return VELOCITY_X;
+}
+
+template<>
+const Variable<double>& GetVelocityVariable<1>()
+{
+    return VELOCITY_Y;
+}
+
+template<>
+const Variable<double>& GetVelocityVariable<2>()
+{
+    return VELOCITY_Z;
+}
+
+template<>
+const Variable<double>& GetShapeVariable<0>()
+{
+    return SHAPE_SENSITIVITY_X;
+}
+
+template<>
+const Variable<double>& GetShapeVariable<1>()
+{
+    return SHAPE_SENSITIVITY_Y;
+}
+
+template<>
+const Variable<double>& GetShapeVariable<2>()
+{
+    return SHAPE_SENSITIVITY_Z;
+}
+}
+
+template <>
+void QSVMSDerivativeUtilities<2>::CalculateStrainRate(
+    Vector& rOutput,
+    const Matrix& rNodalVelocity,
+    const Matrix& rdNdX)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(rOutput.size() != 3)
+        << "rOutput is not properly initialized. [ rOutput.size() != 3, "
+           "rOutput.size() = "
+        << rOutput.size() << " ].\n";
+
+    const IndexType number_of_nodes = rNodalVelocity.size1();
+
+    KRATOS_DEBUG_ERROR_IF(rdNdX.size1() != number_of_nodes)
+        << "rdNdX not initialized properly. [ required rdNdX.size1() = " << number_of_nodes
+        << ", rdNdX.size1() = " << rdNdX.size1() << " ].\n";
+
+    KRATOS_DEBUG_ERROR_IF(rNodalVelocity.size2() != 2)
+        << "rNodalVelocity.size2() != TDim [ TDim = " << 2
+        << ", rNodalVelocity.size2() = " << rNodalVelocity.size2() << " ].\n";
+
+    KRATOS_DEBUG_ERROR_IF(rdNdX.size2() != 2)
+        << "rdNdX.size2() != TDim [ TDim = " << 2
+        << ", rdNdX.size2() = " << rdNdX.size2() << " ].\n";
+
+    noalias(rOutput) = ZeroVector(3);
+    for (IndexType i = 0; i < number_of_nodes; ++i) {
+        rOutput[0] += rdNdX(i, 0) * rNodalVelocity(i, 0);
+        rOutput[1] += rdNdX(i, 1) * rNodalVelocity(i, 1);
+        rOutput[2] += rdNdX(i, 0) * rNodalVelocity(i, 1) + rdNdX(i, 1) * rNodalVelocity(i, 0);
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <>
+void QSVMSDerivativeUtilities<3>::CalculateStrainRate(
+    Vector& rOutput,
+    const Matrix& rNodalVelocity,
+    const Matrix& rdNdX)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(rOutput.size() != 6)
+        << "rOutput is not properly initialized. [ rOutput.size() != 6, "
+           "rOutput.size() = "
+        << rOutput.size() << " ].\n";
```

Comment:
Same, I'd resize.

Reply:
Indeed, you don't need the check as you're initializing below to a `ZeroVector(6)`.

Reply:
AFAIK, I need to since I am using `noalias` 
https://github.com/KratosMultiphysics/Kratos/blob/a8daf70c2019a6ccc997da3144f5639d6da8cc56/applications/FluidDynamicsApplication/custom_elements/data_containers/qs_vms/qs_vms_derivative_utilities.cpp#L138

Reply:
mmmm... yes, I think you're right.